### ISSUE TITLE
AJV Schema should accept useDotenv=false

### ISSUE THREAD

**Issue by jer-sen** [serverless/serverless#8566]
In version 2.14 `configSchema.js`contains `useDotenv: { const: true },` but `false`is a valid value cf https://www.serverless.com/framework/docs/providers/aws/guide/serverless.yml/

Also deprecation warning LOAD_VARIABLES_FROM_ENV_FILES should not be triggered when `useDotenv=false`.

**Comment by medikoo** [comment]
@jer-sen Technically we support only `true`, as we want to remove this setting starting with new major (without ability to turning off `.env` auto load behavior)

In `serverless.yml` we list it to document it, but by convention we also show default setting, `false` is to kind of express  the default (still real default is lack of this setting, technically we could express ith with `null`)

**Comment by jer-sen** [comment]
Ok I will find an other file extension for my environment files.
Anyway, it's quite dangerous to automatically upload `.env` files (often containing sensitive data) by default. Maybe you should reconsider the default behaviour change.

**Comment by medikoo** [comment]
> Ok I will find an other file extension for my environment files.

Why exactly do you have them, what's the purpose?

> Anyway, it's quite dangerous to automatically upload .env files

That's a very good point, it was overlooked that they're packaged, we'll provide a patch shortly

**Comment by jer-sen** [comment]
I use one locally to configure environment variables (loaded with `dotenv ` npm module). But I prefer not to have my production or dangerous (if misconfigured) variables stored and editable on my computer so, since I have only few lambdas, I've manually set my lambda environment variables using AWS console.
I'm open to any idea to better manage all this...

**Comment by medikoo** [comment]
> I use one locally to configure environment variables (loaded with dotenv npm module)

Serverless Framework loads variables only from `.env` files in service folder, and they're not exposed to lambdas (not published to AWS), they're purely for Serverless Node.js process.

**Comment by jer-sen** [comment]
Ok. So there is no security issue 😃 
I'll stay with my workflow, just with a new file extension not used by serverless.

**Comment by takeda** [comment]
@medikoo I use .env with direnv and nix to set up my dev environment.

Direnv takes care of setting up my environment correctly. Since serverless already can use environment variable, this functionality is completely redundant to me. And might possibly cause some weird hard to debug issues.  Please reconsider adding option to disable this and possibly not have it enabled by default.

I think anyone who uses .env/.envrc probably already has some kind of tool that sets environment when entering the directory.

**Comment by medikoo** [comment]
@takeda thanks for input. What exactly issues do you envision? In my understanding worst case scenario is that Framework will just set some environment variables to same values that Direnv written (so will not introduce any effective change)

**Comment by takeda** [comment]
@medikoo it's just the magical behavior that I'm afraid of, since it can create some unexpected behavior that might be harder to trubleshoot.

I'm assuming those variables will be accessible through `${env:...}`.

So for example I have `SOMETHING=1` set in `.env`, and that's what I normally want to use, but I want to run command with different value so I do:

```shell
SOMETHING=2 sls deploy ...
```

But sls uses the value from `.env`.

A variation of it is also, that it uses `.envrc` which is more powerful than `.env`, since it can contain shell statements. By default direnv ignores `.env` unless you place `dotenv` in `.envrc`. I might purposefully remove `dotenv` when testing something and then be surprised that those variables are still accessible.

Most people who use `.env` already use tooling that understand these files. It's good too have programs follow the principle of least astonishment, to avoid strange behavior.

**Comment by medikoo** [comment]
> So for example I have SOMETHING=1 set in .env, and that's what I normally want to use, but I want to run command with different value so I do:
> But sls uses the value from .env.

@takeda Actually`sls` won't override any existing environment variables, with values from `.env` (sorry I've put wrong conclusion in  previous comment.

Check https://github.com/motdotla/dotenv#what-happens-to-environment-variables-that-were-already-set. We follow all recommended practises of `dotenv` package, and do not tweak it's defaults in any way

**Comment by mnapoli** [comment]
To chime in here: I use `.env` with all sorts of frameworks (e.g. Symfony and Laravel) to configure applications running locally. It's a nicer alternative to setting real environment variables for local development.

But on Lambda, we have environment variables and we can set them up in `serverless.yml`, I don't see the point of reading `.env` files. If anything is a bit sensitive, SSM or Secrets Manager is here for that. I'm really afraid of whatever side effects this could have when this magic lands in the next major version (i.e. dev variables being deployed when they shouldn't be).

Maybe this could be an opt-in behavior? E.g. something like:

```yml
service: app

provider:
    environmentFile: .env
```

**Comment by medikoo** [comment]
@mnapoli Serverless doesn't deploy env vars from `.env`, they're only loaded into CLI process (and only if they're not set already). They are not distributed to lambda environments

**Comment by mnapoli** [comment]
@medikoo thanks for the clarification, that sounds much more reasonable then! What's the use case with supporting `.env` btw?

**Comment by medikoo** [comment]
> What's the use case with supporting .env btw?

It's handy to decide which env vars should be used per directory. Without it usually you'd prefix command with needed env vars, which is not very convenient

**Comment by joostdelange** [comment]
Just spend an hour figuring out why my deployed Serverless package did not contain .env files anymore.

It's all okay but it would be cool if this in doc: https://www.serverless.com/framework/docs/providers/aws/guide/packaging the list of default exclude patterns can be updated. Just a heads up 🙂

**Comment by pgrzesik** [comment]
Hello @joostdelange, thanks for reporting it and sorry for the inconvenience, we're already preparing a PR with a fix for that issue: https://github.com/serverless/serverless/pull/8648

**Comment by chekalsky** [comment]
@medikoo This certainly would impact my experience with Serverless as we use `.env` files with Symfony. Symfony has very powerful tooling for `.env` support, e.g. we use `.env.lambda`, `.env.dev.local`, etc. Serverless v3 will provide less powerful solution which will be incompatible with Symfony as I see it. Maybe I am missing something, but for now I don't understand how to have Serverless with Symfony (and probably Laravel too).

**Comment by medikoo** [comment]
> @medikoo This certainly would impact my experience with Serverless

@chekalsky thanks for input. Can outline why exactly this would be problematic?

(Sorry, but I have a notion that concern is just based on assumption, and not on proper understanding of how environment variables resolution from `.env`works in Serverless)

**Comment by chekalsky** [comment]
@medikoo OK, let's try to figure this out. Simple case with Symfony:

`.env`:
```
SENTRY_DSN=https://our-dev-sentry-project
RANDOM_PARAM=true
```

`.env.test`:
```
RANDOM_PARAM=false
```

In this case stage `test` with standard Symfony behaviour will have SENTRY_DSN inherited from `.env`, but in Serverless it will not.

**Comment by medikoo** [comment]
@chekalsky do you run Serverless Framework CLI sometimes through Symfony (which influences environment variables), and sometimes not?

**Comment by chekalsky** [comment]
@medikoo No, I run `sls` manually or via a github action.

**Comment by chekalsky** [comment]
@medikoo btw, what is the right way to pass Cloudformation Outputs into app via serverless? We use environment now:

`MESSENGER_TRANSPORT_DSN: !Ref MessengerQueue`

What will be the new way to do it?

**Comment by medikoo** [comment]
> @medikoo No, I run sls manually or via a github action.

So concern is that `.env` vars which are intended for Symphony project (separate thing from from Serverless service), may be loaded into Serverless CLI process when processing service deployments?

**Comment by medikoo** [comment]
> @medikoo btw, what is the right way to pass Cloudformation Outputs into app via serverless? We use environment now:

For general support please check:
- https://forum.serverless.com/
- https://www.serverless.com/slack
- http://stackoverflow.com/questions/tagged/serverless-framework

**Comment by chekalsky** [comment]
Concern is that `.env` files from Serverless and Symfony will conflict. Serverless will use `.env` file even if it's intended for Symfony and vice versa.

**Comment by medikoo** [comment]
> Concern is that .env files from Serverless and Symfony will conflict. Serverless will use .env file even if it's intended for Symfony and vice versa.

Serverless Framework loads only `.env*` files placed in service folder. Do you have a setup where in same folder you have both Symfony project and Serverless service?

**Comment by chekalsky** [comment]
Yes. `serverless.yml` lies in the root of the project, as in examples.

**Comment by mnapoli** [comment]
From what I understand from the documentation, Serverless v3 will read `.env` files and make variables available to be deployed to Lambda via this technique:

```yaml
service: new-service
provider: aws
  environment:
    SECRET_TOKEN: ${env:SECRET_TOKEN}
```

Is that correct?

Can you confirm that variables defined in `.env` files are not automatically deployed to Lambda but they must be manually included via the `${env:}` syntax above?

If that's the case, then maybe the possibility for an issue is very slim. Because a Symfony/Laravel developer would:

1. either set all variables in `.env` and not use `${env:}` syntax in `serverless.yml` (no problem)
2. or set variables as *real* environment variables (e.g. in CI) and use `${env:}` syntax in `serverless.yml` (no problem)
3. or set all variables in `.env` and use `${env:}` syntax in `serverless.yml` because they know what they are doing and they explicitly want that behavior

Also regarding option `2.` can you confirm that a `.env` variable will never overwrite a real environment variable? (if it's already defined)

**Comment by medikoo** [comment]
> Can you confirm that variables defined in .env files are not automatically deployed to Lambda 

Through `.env` we simply propagate env vars to `process.env` of Serverless Framework CLI process, and that's it.

Putting following to `.env`

```
FOO=bar
```

Is equivalent with running CLI as:

```
$ FOO=bar sls deploy
```

with exception tha settings from `.env` won't override any vars already existing on `process.env`. So having `FOO=bar` in `.env` will have no effect if CLI is invoked as:

```
$ FOO=other sls deploy
```

It's `FOO=other` that will persist.

And the rest works as always worked in a Framework, there were no other changes introduced

**Comment by medikoo** [comment]
> Yes. serverless.yml lies in the root of the project along with the whole project, as in examples.

@chekalsky what examples exactly (?) I don't see any Symphony involved example at https://github.com/serverless/examples/ (but maybe it's name is not  easily discoverable)

Anyway I understand that you feel that loading env vars from `.env` file (in service folder) into Serverless CLI process, in this case, will be harmful. Why exactly?

**Comment by chekalsky** [comment]
(Almost) Any example there includes serverless.yml in the root directory (e.g. https://github.com/serverless/examples/tree/master/aws-node-rest-api-typescript).

What do you mean by service folder? Having a separate folder like `/serverless/serverless.yml`?

**Comment by medikoo** [comment]
@chekalsky _service folder_ is simply a folder in which `serlveress.yml` files lies. It's considered as root of a service. If Serverless Framework CLI is run in context of such folder, then eventual env vars from `.env` file residing in this (service) folder will be propagated onto `process.env`.

My understaning was that you refering to case where same folder is shared as root of both of (1) Serverless service and (2) Symfony project. This use case is still not clear to me, and I don't see it in examples.

Can you provide a specific real world example when loading `.env` files from service folder of Serverless project to Serverless CLI can lead to issues ?

**Comment by mfogel** [comment]
I think the deprecation warning that is printed when serverless encounters a `.env` file now (without `useDotenv: true`  set) is causing some confusion. It could do a better job of explaining what the plan is here.

The current warning (`v2.23.0`) does not mention, and I think probably should:

- variables from `.env` will be loaded into the *build* process only, not automatically distributed to lambda functions
- the setting `useDotenv` is going to be removed altogether in the next major release

Current warning:

> Deprecation warning: Detected ".env" files. Note that Framework now supports loading variables from those files when "useDotenv: true" is set (and that will be the default from next major release)

Suggested possible better warning:

> Deprecation warning: Detected ".env" files. In the next major release variables from ".env" files will be automatically loaded into the serverless build process. Set "useDotenv: true" to adopt that behavior now.

**Comment by medikoo** [comment]
@mfogel that suggestion is very welcome! Can you prepare a PR with that change? We will be very happy to merge it.

**Comment by sergioska** [comment]
I have a compatibility problem with docker-compose. 
docker-compose recognize .env file and load environment variable in "automatic" way, in the same way of serverless. 
So when docker-compose.yml and serverless.yml (and .env file) are in the same path it should be possible disable .env autoload from serverless. 
How can I do to disable that behavior? I try with useDotenv false, but it doesn't seem works!

**Comment by medikoo** [comment]
>How can I do to disable that behavior? I try with useDotenv false, but it doesn't seem works!

Not providing this option, will make Framework not loading `.env` file.

Still from v3 we plan to load `.env` unconditionally, and it's the reason `useDotenv` doesn't support `false` option.

What is exactly the setup on your side?  Why exactly `serverless.yml` and `docker-compose.yml` need to live in same folder?

**Comment by sergioska** [comment]
IMHO I think that possibility to disable this feature could be useful.

**Comment by medikoo** [comment]
> IMHO I think that possibility to disable this feature could be useful.

Can you explain your use case? If final conclusion, that best solution is to provide opt-out, we will definitely provide such possiblity

**Comment by sergioska** [comment]
> Can you explain your use case? If final conclusion would be that best solution is to provide opt-out, we will definitely provide it
I need to use .env in docker-compose, but not in serverless

**Comment by medikoo** [comment]
> I need to use .env in docker-compose, but not in serverless

Why you keep both in same folder, what's the use case?

**Comment by sergioska** [comment]
> > I need to use .env in docker-compose, but not in serverless
> 
> Why you keep both in same folder, what's the use case?

Many projects has docker-compose.yml in project root (such as many project serverless based has serverless.yml in the root of the project)
Anyway I use docker-compose to "dockerize" my application mainly to development and to interacting with some emulated services (for example i use localstack) in local environment.

**Comment by medikoo** [comment]
@sergioska in docker, you may override so it doesn't read `.env` with `--env-file` option -> https://docs.docker.com/compose/environment-variables/#using-the---env-file--option

Anyway, is there a collision between `.env` vars needed for docker compose and for Framework? Is this completely different set of env vars, that you'd prefer to not leak against each other. Or is it about same environment variables which you wish to keep at different values?

**Comment by sergioska** [comment]
Thank you so much for your feedback @medikoo.

> @sergioska in docker, you may override so it doesn't read `.env` with `--env-file` option -> https://docs.docker.com/compose/environment-variables/#using-the---env-file--option
> 

I know it, but this is no good for my use case

> Anyway, is there a collision between `.env` vars needed for docker compose and for Framework? Is this completely different set of env vars, that you'd prefer to not leak against each other. Or is it about same environment variables which you wish to keep at different values?

No, there aren't any collisions (currently). I don't like that serverless load environment variables that it doesn't use and they haven't got significant for its

Anyway, probably exists a workaround or a solution for my use case. 
But I think that possibility to disable this feature in serverless.yml could be useful.

**Comment by pgrzesik** [comment]
> "I don't like that serverless load environment variables that it doesn't use and they haven't got significant for its" 

Hello @sergioska - is there a particular reason as to why you don't like the current behavior? As far as I know, the `docker-compose` exhibits the exact same behavior where it automatically loads `.env` file. 
We would like to understand the use cases before making a decision here, @medikoo also suggested that we might consider adding a CLI flag instead of a configuration property, as it's less about service configuration than about the state of local environment.

**Comment by medikoo** [comment]
Yes, we're open to introduce a support for `--disable-dotenv` CLI param, which will ensure not environment variables from `.env*` files populated in Framework process.

Would that solve the issue for you?

**Comment by Nyholm** [comment]
@medikoo that is a great start. I suggest also making it configurable in serverless.yml.

**Comment by medikoo** [comment]
> @medikoo that is a great start. I suggest also making it configurable in serverless.yml.

We would prefer to not mix it into service configuration, as service configuration is not meant to be CLI specific

We're going into direction where deployment of services can be orchestrated programmatically, and e.g. with one Node.js process it would be possible to deploy many serverless services (so many different `serverless.yml` files), having that, a configuration option which dictates a configuration for Node.js process as a whole, doesn't really fit well a service configuration, and may raise collisions.

**Comment by Nyholm** [comment]
I understand. I see this may bring much value for other users. I **dont** think it is a bad idea and Im **not** trying to stop it. 

However, for my specific use case (and most likely others in this thread). It is useless. 

- I have no benefit that Serverless is reading `.env` files
- I cannot opt-out from this
- It may conflict with values unrelated to Serverless that I've already specified in `.env`
- It removes all my `.env` files before creating the zip archive which will break my production environment

That is why Im would like to see a way to say "This application does not need Serverless to care about my .env files".

There are other suggestions from @chekalsky, @mnapoli, @jer-sen that also make sense. Im no expert in the internals of Serverless, I just wanted to share some input.

**Comment by medikoo** [comment]
> It may conflict with values unrelated to Serverless that I've already specified in .env

Can you elaborate. We haven't found any proof for that. Note that env variables from `.env` are only loaded if they're not already in environment, so technically there's no room for collisions.

We still haven't been provided yet, with a good description of a valid use case, where unconditionally loading `.env` files form  service folder, could be harmful.

Note also that all CLI's I'm aware of, that load `.env` files, do that unconditionally. So providing an opt-out in Serverless CLI, will make a precedence.

**Comment by Nyholm** [comment]
> We haven't found any proof for that. Note that env variables from .env are only loaded if they're not already in environment, so technically there's no room for collisions.

Maybe it is just my ignorance. But I do believe that if I specify `AWS_PROFILE=foobar` in `.env`, it will use the AWS profile "foobar" instead of the profile named "default". Same with other `AWS_*` variables. 


> We still haven't been provided yet, with a good description of a valid use case, where unconditionally loading .env files form service folder, could be harmful.

If you are correct that there are no valid use cases. Then it is just a documentation issue to address my ignorance/fear. 

> Note also that all CLI's I'm aware of, that load .env files, do that unconditionally. So providing an opt-out in Serverless CLI, will make a precedence.

Most other CLI's dont copy all files in the current folder and upload it on S3. And the ones that do (AWS CLI) dont read `.env`. 
But I've have relative little knowledge about other tools. I may be wrong.

**Comment by medikoo** [comment]
> Maybe it is just my ignorance. But I do believe that if I specify AWS_PROFILE=foobar in .env, it will use the AWS profile "foobar" instead of the profile named "default". Same with other AWS_* variables.

Yes, in this specific case, if someone relies on _default_ profile fallback, it's true. Still can you describe a valid use case for such configuration (?) Where in a Serverless service folder, there's `.env` file with `AWS_PROFILE` setup, which by no means should be used by `serverless` process, as user decided to use _default_ profile in such case.

I know that in theory we can throw hundreds of cases, problem is that we still haven't haven't heard a single valid real world use case that will validate the opt-out option.

**Comment by Nyholm** [comment]
Let me describe my current setup: 

I deploy with CI from gitlab. The jobs starts like this: 

``` 
  before_script:
    - mkdir ~/.aws
    - printf "[profile default]\nregion = eu-central-1" >>  ~/.aws/config
    - printf "[default]\naws_access_key_id = $AWS_KEY_ID\naws_secret_access_key = $AWS_SECRET" >>  ~/.aws/credentials
```

It will add some AWS credentials to a profile named "default" into the docker container that the job runs in. 

Then I'll deploy my application to AWS using the following command. This will use the AWS profile named "default". 

```
serverless deploy --stage=prod --force
```

---------

If my PHP application needs the environment variable `AWS_PROFILE`, I would specify it in `.env`. Best case scenario, the deployment CI job fails. Worst case scenario, the new value of `AWS_PROFILE` matches another profile defined in my CI job...

Why the PHP application needs `AWS_PROFILE`? There could be 100 reasons. Maybe I have setup my AWS account years ago and I've not migrated all the config to Serverless yet or the **A**utomatic **W**ork **S**ystem needs config...

Since the CI job config is now connected to both the application config and the version of Serverless, there must be a coordination between developers and devops every time CI is changing, `.env` is changing or Serverless is updated. 

Will I manage? Yes, I'll probably find workarounds (like #9484). I could also specify the `--aws-profile` option when I run `deploy`.... but again, this is all for a feature I don't care about and that I can't disable. 

----------------

Could I suggest that Serverless won't read the `/my/project/.env` file, but instead read `/my/project/.serverless/.env`?

That would avoid all conflicts will existing projects and still add the functionality to read environment from file.

**Comment by medikoo** [comment]
@Nyholm thanks for description. What is "PHP Application" ? Why it shares the folder with Serverless service? Which do you feel is more _cardinal_ in this case? Is this that Serverless service complements the "PHP Application", or is it the other way? 
Can you describe briefly what "PHP Application" does and what "Serverless service" does?

**Comment by Nyholm** [comment]
> What is "PHP Application" ?

Any PHP application. In my specific case it is most likely Symfony. This is my application, my website, my API etc. 

> Why it shares the folder with Serverless service?

Maybe I misunderstand you here. My website lives in `/my/project`. Im always referring to the `/my/project/.env` when I write "`.env`". I assume that you refer to my application/website as a "Serverless service". 

>  Which do you feel is more cardinal in this case? Is this that Serverless service complements the "PHP Application", or is it the other way?

It is the PHP application/website that is the main thing for me. Serverless is just a tool. 

> Can you describe briefly what "PHP Application" does and what "Serverless service" does?

I think we are using different terminology. Maybe you can elaborate based on my this message.

**Comment by medikoo** [comment]
> Maybe I misunderstand you here. My website lives in /my/project. Im always referring to the /my/project/.env when I write ".env". I assume that you refer to my application/website as a "Serverless service".

Issue you're describing only happens if your application/website is configured in same folder as Serverless service (so for both application root folder and serverless service root folder is same).

Is this the case on your side? Do you keep your Serverless service (so `serverless.yml` file) in same folder in which you configure your application/website? If so, why?

**Comment by Nyholm** [comment]
Yes. It is true that my application lives in `/my/project`. Here you find `.env`, `index.php` and `serverless.yml`. 

> If so, why?

That is an excellent question. =)

My application lives there because that is basically how you do it in PHP. One application per git repository and the application root is the same as the git root. 

Im not sure why `serverless.yaml` is in the root too, where else should I put it?

**Comment by medikoo** [comment]
> Im not sure why serverless.yaml is in the root too, where else should I put it?

Simply keep PHP project and Serverless service at two different folders, especially if there are no common parts between them.

Mixing two unrelated applications within one folder feels as bad pattern to me. It's hard to upfront assume which file belongs to which, and that makes it way harder to reason about them.

**Comment by Nyholm** [comment]
> Simply keep PHP project and Serverless service at two different folders, especially if there are no common parts between them.

Interesting. In my PHP focused mind, this does not make any sense for me. This is a good example of the root files on a normal PHP website deployed with serverless. As you see, there is all kinds of config files there that belongs to different tools and package managers. 

<img width="279" alt="Screenshot 2021-05-17 at 16 45 40" src="https://user-images.githubusercontent.com/1275206/118508492-89e89880-b72f-11eb-888c-82ea25eae214.png">

I know Im not an exception here. =)

> Mixing two unrelated applications within one folder feels as bad pattern to me.

That is not the way I see it. I see my `serverless.yaml` as a configuration file to the `serverless` tool I use to deploy my application. Im happy to use `.serverless/serverless.yml`, but I do see that they live in the same git repository (and git root is the same as PHP project root). 

> ...  that makes it way harder to reason about them

I think that is the situation we are in. You may convince me that this is a bad idea and I'll maybe change. But there are a lot of other PHP developers that wont. Hence the discussion about Serverless reading `.env` files.

**Comment by mnapoli** [comment]
I think @Nyholm is trying to say that `serverless.yml` lives at the root of the project directory because that's how the project is deployed.

Just like you have a README, or a Makefile, or a `docker-compose.yml` file at the root of the project.

> Mixing two unrelated applications within one folder

In that context, they are not really unrelated: the PHP project is the one being deployed to Lambda. There is only 1 application here.

Does that make more sense?

Let's say the project is an HTTP API (whatever the language). If `serverless.yml` deploys the API to AWS, where would you put `serverless.yml` @medikoo?

If I look at official Serverless examples, the `serverless.yml` file is always at the root of the GitHub project, for example: https://github.com/serverless/examples/tree/master/aws-node-twitter-joke-bot

**Comment by medikoo** [comment]
> In that context, they are not really unrelated: the PHP project is the one being deployed to Lambda

Thanks @mnapoli for that clarification. It was totally not obvious to me, I assumed both are not that related.

In that case, the PHP project is part of Serverless Service, and I believe right way to approach it, is to put PHP project  to some dedicated folder. Ideally code that's used to deploy the app should not be mixed with code that's run in the cloud.

> If I look at official Serverless examples, the serverless.yml file is always at the root of the GitHub project, for example: https://github.com/serverless/examples/tree/master/aws-node-twitter-joke-bot

Yes, if repository hosts just single Serverless service it's totally natural and right to have `serverless.yml` at root (it would be weird to set it up differently), but if repository would host multiple Serverless services, or Serveless service and e.g. Ruby on rails application, I would definitely nest each entity in its own folder and not try to mix code that's not common to all of them

**Comment by Nyholm** [comment]
> In that case, the PHP project is part of Serverless Service, and I believe right way to approach it, is to put PHP project to some dedicated folder.

But that not how people do PHP projects =)

> Ideally code that's used to deploy the app should not be mixed with code that's run in the cloud (especially it if it implies a risk of conflicts)

True. But there is no "**code** to deploy the app". It is just a config file (serverless.yml)

We (PHP people) are aware of the "risk of conflicts". That is why we are desperate to fin a way to disable the new "read env files" feature. 

------------

I believe this input shows a bit the situation this new Serverless feature has put us in. I hope there could be a workaround. 

Ie Make the name of `.env` that Serverless reads configurable. Or use `/my/project/.serverless/.env` or to add an option to disable this feature.

**Comment by medikoo** [comment]
> I believe this input shows a bit the situation this new Serverless feature has put us in.

From where I stand, this is just one example of collisions, which may happen when we try to maintain two different projects in same folder.

Technically we can fix this specific collision with some opt-out workaround, but with that approach we won't fix any other collisions which may occur due to this kind of organization. The only solid fix is to move PHP project into other folder.

**Comment by chekalsky** [comment]
To continue the conversation started — why do you consider serverless as different application and not the part of what's in the git root? I always was thinking about `serverless.yml` as of `package.json` or `docker-compose.yml` which describes how application should be deployed, not as another application itself.

Would it be feasible to provide an opt-out option for this new dotEnv behaviour?

**Comment by christian-eriksson** [comment]
I wrote this in the forum thread asking about adding a `useDotenv=false` option, but this might be a better place.

One example of a use case where I might want to set useDotenv to false is if I’m using `serverless.ts` config files. Where I prefer loading my variables in my own modules and importing them into the Serverless config. In my config modules I might do some checking and verification on the environment. 

This environment may be loaded from dotenv files. This means that the loading and verification of the environment is done before Serverless actually gets involved. It may also be that my application doesn't load dotenv files under the same rules as Serverless, perhaps I load a different dotenv than Serverless.

Basically this means that my environment, once Serverless actually starts deploying (or my application starts locally), may be different than when I verified it. Which to me seems like a potentially huge problem.

**Comment by christian-eriksson** [comment]
To be clear a CLI flag, like the one @medikoo suggested would solve this, I see no need to necessarily put the dotenv behavior  config in the serverless file.

**Comment by medikoo** [comment]
@chekalsky we're going into direction where Serverless services could be deployed programmatically, e.g. you would be able to require a `Serverless` instance and deploy programmatically multiple services at once. In that model handling `useDotenv` configuration at service level cannot be done reliably, as environment variables affects whole Node.js process, and not single service in such case (imagine one service having `useDotenv: false` and other `useDotenv: true`).

So summarizing, technically it's a CLI process specific setting, and not a Serverless service specific setting that's why I believe it doesn't belong to service configuration, and we propose to introduce opt-out via CLI param

Similarly is with `package.json`, in one process you can handle multiple of those, and none of them should affect global state of a process.

**Comment by siolfyr** [comment]
This feels like a mistake that will cause a lot of problems for users. Is the benefit of adding this so great that causing all of those problems is worth it? Why purposely use a name which conflicts with programs previously designed using Serverless?

I don't think it should be opt-out, but rather opt-in.

**Comment by takeda** [comment]
My feeling exactly, despite all of the concerns it still being pushed. And not only it is enabled by default, there's no option to even disable it. I already have a tool that uses `.env` file and I'm perfectly happy with how it works.

What I'm trying to avoid is wasting few hours on an issue that could be traced to this behavior. At very least use a different name like `.serverlessrc`

**Comment by pgrzesik** [comment]
Hello @siolfyr and @takeda - thanks for voicing your opinion here. We are going to introduce a CLI option that will allow to opt-out from this behavior, as @medikoo mentioned in one of the previous comments. Would that satisfy your needs? What potential issues do you see with this behavior? Maybe there's a case that we're missing that is really problematic

**Comment by medikoo** [comment]
We've decided (internally) to revert from deprecation notice, and in v3 we will keep the very same behavior as it's in v2 (`.env` files will be loaded by the Framework only if `useDotenv: true` setting is set).

The reasoning is that we're worried to break for users which already use `.env` files to configure the environment for lambdas, where the same root folder is shared by both service configuration and lambda logic

PR that removes the deprecation in v2 will be introduced in the next days (once it's in, we will close this issue)

**Comment by chekalsky** [comment]
❤️ Thank you

**Comment by ljwagerfield** [comment]
Thank you @medikoo. A war story FWIW...

For reasons I'm still unsure of, setting `useDotenv=true` caused our live environments to fail due to our EC2 instances listening on the wrong port: the port they were listening on came from the `.env` file. Whilst I can't seem to recreate the problem locally (the artifacts that I build locally do indeed have the correct port in them), all I can say is: the problem in CI is caused by having `useDotenv=true`.

**Comment by medikoo** [comment]
@ljwagerfield all that `useDotenv: true` does, is loading (not yet populated) environment variables from `.env` (or `.env<stage>`), that's it.

**Comment by ljwagerfield** [comment]
> loading (**not yet populated**) environment variables from .env (or .env<stage>), that's it.

This would cause an issue if the desired env vars (being set elsewhere, say by another plugin) followed the same logic. For example, if they were only being set when absent, then they will no-longer be set after setting `useDotenv=true`. I'm not sure if that's the exact issue I'm facing or not, but just raising as it's not as innocent of a behaviour as it may seem.

**Comment by ljwagerfield** [comment]
> all `useDotenv: true` does is load (not yet populated) environment variables from .env (or .env), that's it.

OK, so this is the problem I'm encountering:

> This would cause an issue if the desired env vars (being set elsewhere, say by another plugin) followed the same logic.

The specifics are:

1. Our `.env` file contains local developer-only config, not intended to be used for deployments.
2. We use [`serverless-plugin-scripts`](https://www.serverless.com/plugins/serverless-plugin-scripts) to run a post-deployment action from serverless. It calls `npm run deploy`.
3. `npm run deploy` calls [`serverless-export-env`](https://www.serverless.com/plugins/serverless-export-env) to extract  the `environment` section of `serverless.yml` into some temporary file (`temp.env`).
4. `npm run deploy` then loads these env vars using [`dotenv`](https://github.com/motdotla/dotenv)
5. `npm run deploy` then calls `webpack` which builds a bundle. It uses the `EnvironmentPlugin`.

When setting `useDotenv: true` the `.env` file (containing local vars) is loaded into the `serverless` process, and when `serverless` spawns the `npm run deploy` command via `serverless-plugin-scripts` it inherits those environment vars from the parent `serverless` process. As such, when step 4 runs, it doesn't load the intended PROD vars, as the LOCAL vars are already loaded from step 1.

**Comment by medikoo** [comment]
> Our .env file contains local developer-only config, not intended to be used for deployments.

@ljwagerfield then you should not use `useDotenv: true` setting. This trigger is literally about loading environment variables from `.env` files for deployment specifically (it's not to populate environment variables into Lambda config).

**Comment by ljwagerfield** [comment]
Thank you @medikoo, glad it won't be mandatory in v3 🙏
